storage/storage.go

```diff
@@ -31,22 +32,211 @@ func NewPaginationOptions(ps int32, contToken string) PaginationOptions {
 	}
 }
 
-// TupleIterator is an iterator for Tuples. It is closed by explicitly calling Stop() or by calling Next() until it
-// returns an TupleIteratorDone error.
-type TupleIterator interface {
-	Next() (*openfgapb.Tuple, error)
-	// Stop will release any resources held by the iterator. It must be safe to be called multiple times.
+type Iterator[T any] interface {
+	Next() (T, error)
 	Stop()
 }
 
 // ObjectIterator is an iterator for Objects (type + id). It is closed by explicitly calling Stop() or by calling Next() until it
-// returns an ObjectIteratorDone error.
-type ObjectIterator interface {
-	Next() (string, error)
-	// Stop will release any resources held by the iterator. It must be safe to be called multiple times.
-	Stop()
+// returns an IteratorDone error.
+type ObjectIterator = Iterator[*openfgapb.Object]
+
+// TupleIterator is an iterator for Tuples. It is closed by explicitly calling Stop() or by calling Next() until it
+// returns an IteratorDone error.
+type TupleIterator = Iterator[*openfgapb.Tuple]
+
+// TupleKeyIterator is an iterator for TupleKeys. It is closed by explicitly calling Stop() or by calling Next() until it
+// returns an IteratorDone error.
+type TupleKeyIterator = Iterator[*openfgapb.TupleKey]
+
+// UniqueObjectIterator takes two ObjectIterators and yields an ObjectIterator that returns only distinct objects with the
+// duplicates removed.
+func UniqueObjectIterator(iter1, iter2 ObjectIterator) (ObjectIterator, error) {
+	objectMap := map[string]*openfgapb.Object{}
+
+	for {
+		val, err := iter1.Next()
+		if err != nil {
+			if errors.Is(err, IteratorDone) {
+				break
+			}
+			return nil, err
+		}
+
+		objectMap[tuple.ObjectKey(val)] = val
+	}
+
+	for {
+		val, err := iter2.Next()
+		if err != nil {
+			if errors.Is(err, IteratorDone) {
+				break
+			}
+			return nil, err
+		}
+
+		objectMap[tuple.ObjectKey(val)] = val
+	}
+
+	iter := &channelledStaticIterator[*openfgapb.Object]{
```

Comment:
We use a channelledStaticIterator here instead of additional allocations for `[]*openfgapb.Object`. The overhead for a buffered channel that yields the values from the map is lower cost than building a new slice of the objects and then returning a static iterator over the new slice.

Reply:
We are already doing more allocations than we need here by retaining the objects in the maps.

Reply:
But those allocations are separate from the ones I was referring to with this optimization. The allocations to the `objectMap` are necessary to build the distinct set between the two iterators. I'm not sure how we'd eliminate those allocations unless we had a clever iterator that somehow new what it already yielded without storing anything in memory. 

I just benchmarked the difference between this and doing a copy of the values in the map into a slice and then returning a static iterator over the slice and the it's actually quite a bit faster. So the map value copy and then static iterator is better than this. I'll just drop the `channelledStaticIterator` then 👍

Reply:
Turns out the memory copy to a slice has less overhead than the bandwidth of the channel send/receive. So it's better overall.